### ISSUE TITLE
Performance problems

### ISSUE THREAD

**Issue by nkiesel** [gluster/gluster-kubernetes#410]
We are trying to replace an (inflexible) NFS-based storage solution with gluster-kubernetes.  While we got that working, we see a 100% - 300% slowdown.  Is that to be expected?

**Comment by JohnStrunk** [comment]
First thing to check is your network bandwidth. Gluster replicates from the
clients, so writing at 33MB/s will use 100MB/s of network BW ~1GbE.
(Assuming 3 way replication)

**Comment by JohnStrunk** [comment]
@nkiesel Are you still having unexpected performance or can this issue be closed?

**Comment by nkiesel** [comment]
We still face the same issues.  I briefly checked our network and the testbed is using some vmware virtual machines for running a Kubernetes cluster.  They never gave us network perf problems before, but I will measure to be sure.  Would a simple `nc` client/server test be sufficient?

**Comment by JohnStrunk** [comment]
For network throughput, I'm a bigger fan of ttcp:
* Receiver: `ttcp -rs -n 102400`
* Sender: `ttcp -ts -n 102400 dest_host`
That'll send ~800 MB between the hosts and give you the calculated BW.

Some other ideas...
* See what happens w/ a volume created from a single brick (i.e., no replication). Obviously don't use that in production, but you can at least see what BW you get.
* Take a look at your workload... Gluster tends to struggle w/ lots of small files. Try fio w/ large sequential I/O and see what you get.

**Comment by martinsimon007** [comment]
Sorry for the delay, as we were facing some issues while upgrading our kubespray cluster.

### Scenario 1-- Between glusterfs pods on different nodes
- **Reciever statistics**
`ttcp-r: buflen=8192, nbuf=102400, align=16384/0, port=5001  tcp`
`ttcp-r: socket`
`ttcp-r: accept from 172.27.141.192`
`ttcp-r: 838860800 bytes in 1.54 real seconds = 0.51 GB/sec +++`
`ttcp-r: 104491 I/O calls, msec/call = 0.02, calls/sec = 67933.57`
`ttcp-r: 0.0user 1.2sys 0:01real 84% 0i+0d 630maxrss 0+2pf 4537+102csw`
- **Transmitter statistics**
`ttcp-t: buflen=8192, nbuf=102400, align=16384/0, port=5001  tcp  -> 172.27.141.191`
`ttcp-t: socket`
`ttcp-t: connect`
`ttcp-t: 838860800 bytes in 1.53 real seconds = 0.51 GB/sec +++`
`ttcp-t: 102400 I/O calls, msec/call = 0.02, calls/sec = 66880.28`
`ttcp-t: 0.0user 0.2sys 0:01real 17% 0i+0d 598maxrss 0+2pf 507+4csw`

### Scenario 2 -- Between tomcat and glusterfs pods on different nodes
- **Receiver (Tomcat) statistics**
  `ttcp-r: buflen=8192, nbuf=102400, align=16384/0, port=5001`
`ttcp-r: sockbufsndsize=16384, sockbufrcvsize=87380, sockbufsize=51882,  # tcp #`
`ttcp-r: accept from 10.233.75.0`
`ttcp-r: 838860800 bytes in 2.095 real seconds = 0.373 GB/sec +++`
`ttcp-r: 104070 I/O calls, msec/call = 0.021, calls/sec = 49674.611`
`ttcp-r: 0.012user 1.280sys 0:02real 61% 0i+0d 790maxrss 0+2pf 7455+52csw`

- **Transmitter (Gluster) statistics**
  `ttcp-t: buflen=8192, nbuf=102400, align=16384/0, port=5001  tcp  -> 10.233.71.7`
`ttcp-t: socket`
`ttcp-t: connect`
`ttcp-t: 838860800 bytes in 2.09 real seconds = 0.37 GB/sec +++`
`ttcp-t: 102400 I/O calls, msec/call = 0.02, calls/sec = 49035.24`
`ttcp-t: 0.0user 0.5sys 0:02real 27% 0i+0d 600maxrss 0+2pf 448+7csw`

### Volume information
The volumes are distributed volumes with a single brick per volume. Given below is the volume information for the tomcat PV

`Volume Name: vol_14707df8e74dd4727819b795d7a8c1bc`
`Type: Distribute`
`Volume ID: 0fbfc3cd-5966-46e4-a8e3-76f5d18e6846`
`Status: Started`
`Snapshot Count: 0`
`Number of Bricks: 1`
`Transport-type: tcp`
`Bricks:`
`Brick1: 172.27.141.193:/var/lib/heketi/mounts/vg_9e7202bbde7f2a15137571663af955c7/brick_c2d0d075d3a6a8ec9be03e69e4e5b238/brick`
`Options Reconfigured:`
`transport.address-family: inet`
`nfs.disable: on`
`cluster.brick-multiplex: on`

**Comment by JohnStrunk** [comment]
I'm assuming @martinsimon007 and @nkiesel work together? If not, lets split this into two different issues.

Going on the assumption this is the same system...
So we've got ~510 MB/s between gluster pods and ~370 MB/s between gluster and an app pod. I expect the difference between these two to be the overhead of kube's overlay network. Gluster uses the host network, so the gluster-to-gluster path is using the node's interfaces directly. Whereas the Gluster to Tomcat requires the packets use kube's network.

* What was your storage performance before (and to what sort of NFS server)?
* What is the performance to the gluster volume, above?
* What is the hardware you are using for gluster? Node specs and storage system specs (HDD, SSD, locally attached, carved out of a VMWare storage pool, etc.).

**Comment by nkiesel** [comment]
yeah, same team.  Sorry we did not make that clear. Martin will be
providing the data you asked for.  Thanks for trying to help us!

**Comment by martinsimon007** [comment]
## VM Details
**OS** : Ubuntu
**Browser** : Firefox 56.0
**GlusterFS** : Distributed with 1 brick per volume
**Kubernetes** : kubespray (kubernetes version 1.7.3)

**Kubespray cluster** : Three VMs (VM1, VM2, VM3) 
**GlusterFS cluster** : Same VMs as above with one disk attached to each (VM1, VM2, VM3)
**NFS** :  The nfs-provisioner for kubernetes present at the below URL is used
[https://github.com/kubernetes-incubator/external-storage/tree/master/nfs](url)
It uses a single disk attached to VM1 

### Application Stack
**UI** : Web application written in backbone
**Application Server** : Tomcat (Runs as a pod)
**Datastore** : Oracle (Runs as a pod)

All datafiles for the web application and the datastore are in PVs

## Storage disk details
### NFS

VM NAME | Partition | Disk/Partion Type | Type OF HDD | Rotational Speed | SAS Interface Access Speed | Single HDD Capacity
-- | -- | -- | -- | -- | -- | --
VM1(NFS) | /dev/mapper/vmidoctest6--vg-root | Logical Partion | SAS | 15 K RPM(Spindle Speed) | 6 GB | 600 GB

### GlusterFS

VM NAME | Partition | Disk/Partion Type | Type OF HDD | Rotational Speed | SAS Interface Access Speed | Single HDD Capacity
-- | -- | -- | -- | -- | -- | --
VM1 (GlusterFS) | /dev/sdb | Logical Extended Partion(LVM extended) | SAS | 15 K RPM(Spindle Speed) | 6 GB | 100 GB
VM2 (GlusterFS) | /dev/sdb | Logical Extended Partion(LVM extended) | SAS | 15 K RPM(Spindle Speed) | 6 GB | 100 GB
VM3 (GlusterFS) | /dev/sdb | Logical Extended Partion(LVM extended) | SAS | 15 K RPM(Spindle Speed) | 6 GB | 100 GB

## Performance Comparison
Storage Class | Tomcat pod startup | Tomcat pod restart
-- | -- | --
NFS | 1 min 44 secs | 1 min 30 secs
Gluster | 6 min 49 secs | 6 min 44 secs


Storage Class | Login to web application | Open application form | Submit application form | Open report
-- | -- | -- | -- | --
NFS | 9.42 secs | 20.41 secs | 1.89 secs | 5.50 secs
Gluster | 23.11secs | 42.91 secs | 3.57 secs | 16.23 secs

### Application terminology
**Form** : A page where user input is captured, which on submission writes to the datastore
**Report** : A page with a tabular report populated with data obtained from the datastore

**Comment by nelsonfassis** [comment]
Was there any progress in this issue?  I am also trying to improve performance on my Gluster cluster

**Comment by nkiesel** [comment]
Not really, though at least partially because our team shifted focus and
did not interacted enough with the Gluster team.  We still have the
weirdness that Tomcat startup is 3 times slower when running on top of
glusterfs, but that we only see a small performance hit once tomcat is up
and running.  I suspected excessive classpath scanning but that was never
proven.

BTW: we also tried portworx and that was fast and reliable for us.
However, remember that portworx is a commercial offering.
